docs/writing_tests.md

```diff
@@ -0,0 +1,156 @@
+        cgh.single_task<class simple_kernel>([=] {
+            w[0] = 42;
+        });
+    });
+
+    sycl::host_accessor r{buf, sycl::read_only};
+    CHECK(r[0] == 42);
+}
+```
+
+This adds a test case with the description `"a simple test case"` and the *tag* `[simple]`.
+Both can later be used to narrow down the set of test cases that will be executed during runtime.
+
+When configuring CMake, the new test category will automatically be detected and a target with the name `test_simple` is added.
+You can run the test case by either executing `./bin/test_simple` directly, or alternatively as part of `./bin/test_all`.
+
+> Note that for historic reasons, the CTS currently contains many test cases that are written in a different style.
+> Please see [New-style vs Legacy Test Cases](#new-style-vs-legacy-test-cases) for more information.
+
+## Important Catch2 Concepts
+
+The SYCL CTS relies on [Catch2](https://github.com/catchorg/Catch2/) as its underlying testing framework.
+This section will list the most important concepts required to write tests with Catch2.
+For a comprehensive overview of all features, please refer to the [Catch2 documentation](https://github.com/catchorg/Catch2/tree/devel/docs).
+In addition, the CTS provides several custom utilities to extend Catch2's feature set.
+See [Special Macros & Custom Matchers](#special-macros--custom-matchers) for more information.
+
+### Test Case Macros
+
+Catch2 provides several macros of varying complexity for defining test cases.
+While different macros take different parameters, they all require a *description* and optionally a list of *tags* to be specified.
+
+
+- `TEST_CASE`: This is the most basic macro, useful for test cases that do not require multiple compile-time definitions.
+- `TEMPLATE_TEST_CASE`
+- `TEMPLATE_TEST_CASE_SIG`
+
+**TODO**
+
+### Sections
+
+**TODO**
+
+### Generators
+
+**TODO**
+
+## Special Macros & Custom Matchers
+
+**TODO**
+
+## Best Practices
+
+Here is a list of best practices for writing test cases.
+These are not set in stone and are likely to evolve over time.
+
+- Always write tests using Catch2 macros, avoid [legacy test cases](#new-style-vs-legacy-test-cases).
+- Avoid old-style `if(!condition) FAIL("reason");` pattern.
+  Use `CHECK(condition)` instead.
+- Keep test cases small and focused to a single concept / behavior.
+  Even a single function could be tested with several test cases.
+- Use natural language descriptions for test cases.
+    - Avoid: `"host_accessor range mismatch exception"`.
+    - Prefer: `"host_accessors throws if accessed range exceeds buffer dimensions"`.
+- Tag test cases according to the feature being tested:
+    - Use `[some_type]` for types that exist in the SYCL specification (example: `[host_accessor]`).
+    - Use `[some-concept]` for concepts without a clearly associated type (example: `[backend-interop]`).
+- Try to write test cases in the same order as the associated API specification (if possible).
```

Comment:
I think it would be good to provide some guidance on how many test cases to include in a single source file.  For example, is it appropriate to have a single source file that has dozens of test cases for a single SYCL API (e.g. there will probably be a lot of test cases for `accessor`)?  Or, is it more appropriate to create a test category for `accessor` and have many source files?  If this is the case, should each source file have just one test case, or should related ones be grouped into the same source file?

Reply:
Agreed, although I don't have a clear answer here yet. I'm leaning towards fewer files with multiple test cases each, though. The only reasons for having multiple files at all (that I can think of right now) would be to group them conceptually, and to keep incremental build times low.